### PR TITLE
llvm: fix for Flang variant due to exception handling

### PR DESCRIPTION
**Pull request by pmccormick**
In the main llvm-project repo Flang will not allow LLVM_ENABLE_EH to be set (the project developer guidelines explicitly call out that exceptions should not be used).  This PR adds a check to make sure LLVM_ENABLE_EH is not set when the Flang variant is added.  Without it, Flang builds off the main LLVM branch and future releases (likely after 16.x) will fail.

### CODE REVIEW

File: var/spack/repos/builtin/packages/llvm/package.py [hunk 1]
```diff
@@ -545,8 +545,20 @@ def cmake_args(self):
...
+        # LLVM_ENABLE_EH=True when building flang will soon
+        # fail (with changes at the llvm-project level).
+        # Only enable exceptions in LLVM if we are *not*
+        # building flang.  FYI: LLVM <= 16.x will build flang
+        # successfully but the executable will suffer from
+        # link errors looking for C++ EH support.
+        if "+flang" not in spec:
+            cmake_args.append(define("LLVM_ENABLE_EH", True))
+
+        cmake_args = cmake_args + [
```
Comment (haampie) [inline var/spack/repos/builtin/packages/llvm/package.py]:
```suggestion
        cmake_args.extend([
```

**Comment by haampie** [comment]
Do you have a reference for this? A link in the comment would help

**Comment by pmccormick** [comment]
> Do you have a reference for this? A link in the comment would help

The associated LLVM upstream patch was accepted and has landed in the main branch.   

See:  https://reviews.llvm.org/D146173

**Comment by balay** [comment]
(with current xsdk/develop branch)
```
[balay@pj01 spack]$ ./bin/spack install llvm@main+flang

<snip>

1 error found in build log:
     351    -- Performing Test C_SUPPORTS_WERROR_GLOBAL_CONSTRUCTOR
     352    -- Performing Test C_SUPPORTS_WERROR_GLOBAL_CONSTRUCTOR - Failed
     353    -- Performing Test CXX_SUPPORTS_WERROR_GLOBAL_CONSTRUCTOR
     354    -- Performing Test CXX_SUPPORTS_WERROR_GLOBAL_CONSTRUCTOR - Failed
     355    -- Performing Test COMPILER_SUPPORTS_WARNING_WEAK_VTABLES
     356    -- Performing Test COMPILER_SUPPORTS_WARNING_WEAK_VTABLES - Failed
  >> 357    CMake Error at /home/balay/spack/spack-stage/spack-stage-llvm-main-ta4eua5lfv4jly6n22gdrnfd4syanhnf/spack-src/flang/CMakeLists.txt:19 (message):
     358      Flang does not currently support building with LLVM exceptions enabled.
     359      Please disable LLVM_ENABLE_EH when building flang.
     360
     361
     362    -- Configuring incomplete, errors occurred!
```

File: var/spack/repos/builtin/packages/llvm/package.py [hunk 2]
```diff
@@ -559,8 +559,20 @@ def cmake_args(self):
...
+        # LLVM_ENABLE_EH=True when building flang will soon
+        # fail (with changes at the llvm-project level).
+        # Only enable exceptions in LLVM if we are *not*
+        # building flang.  FYI: LLVM <= 16.x will build flang
+        # successfully but the executable will suffer from
+        # link errors looking for C++ EH support.
+        if "+flang" not in spec:
+            cmake_args.append(define("LLVM_ENABLE_EH", True))
+
+        cmake_args.extend([
```
Comment (balay) [inline var/spack/repos/builtin/packages/llvm/package.py]: https://github.com/spack/spack/actions/runs/4630496708/jobs/8192146288?pr=36171

`var/spack/repos/builtin/packages/llvm/package.py:575: [E999] SyntaxError: '(' was never closed`


Perhaps you can  simplify this change to:
```patch
diff --git a/var/spack/repos/builtin/packages/llvm/package.py b/var/spack/repos/builtin/packages/llvm/package.py
index f62822e3ca..3030f01cc8 100644
--- a/var/spack/repos/builtin/packages/llvm/package.py
+++ b/var/spack/repos/builtin/packages/llvm/package.py
@@ -555,7 +555,6 @@ def cmake_args(self):
         cmake_args = [
             define("LLVM_REQUIRES_RTTI", True),
             define("LLVM_ENABLE_RTTI", True),
-            define("LLVM_ENABLE_EH", True),
             define("LLVM_ENABLE_LIBXML2", False),
             define("CLANG_DEFAULT_OPENMP_RUNTIME", "libomp"),
             define("PYTHON_EXECUTABLE", python.command.path),
@@ -564,6 +563,16 @@ def cmake_args(self):
             from_variant("LLVM_ENABLE_ZSTD", "zstd"),
         ]
 
+        # Flang does not support exceptions from core llvm.
+        # LLVM_ENABLE_EH=True when building flang will soon
+        # fail (with changes at the llvm-project level).
+        # Only enable exceptions in LLVM if we are *not*
+        # building flang.  FYI: LLVM <= 16.x will build flang
+        # successfully but the executable will suffer from
+        # link errors looking for C++ EH support.
+        if "+flang" not in spec:
+            cmake_args.append(define("LLVM_ENABLE_EH", True))
+
         version_suffix = spec.variants["version_suffix"].value
         if version_suffix != "none":
             cmake_args.append(define("LLVM_VERSION_SUFFIX", version_suffix))
```

Comment (balay) [reply var/spack/repos/builtin/packages/llvm/package.py]: Would be good if this issue is resolved

**Comment by naromero77** [comment]
@pmccormick It is not passing checks, can you take look?

File: var/spack/repos/builtin/packages/llvm/package.py (hunk 2 above)
Comment (balay) [reply var/spack/repos/builtin/packages/llvm/package.py]: Pushed this suggested fix to this PR/branch

**Review by balay — APPROVED** [review]

**Comment by naromero77** [comment]
Its a minor change, but want to give the others a chance to chime in @trws @haampie.

**Pull request merged**